src/java/org/apache/cassandra/nodes/Nodes.java

```diff
@@ -18,443 +18,924 @@
+ * <li>time to serialize 1000 {@link PeerInfo}: ~25ms</li>
+ * <li>time to update {@link LocalInfo} in a blocking fashion w/ synchronous persistence: ~90µs</li>
+ * <li>time to update {@link LocalInfo}, mark as dirty and trigger async persistence: ~400ns</li>
+ * <li>time to update one {@link PeerInfo}, mark as dirty and trigger async persistence: ~400ns</li>
+ * </ul>
  */
 public class Nodes
 {
     private static final Logger logger = LoggerFactory.getLogger(Nodes.class);
 
-    @VisibleForTesting
     private final ExecutorService updateExecutor;
 
-    private final INodesPersistence nodesPersistence;
+    private final Path baseDirectory;
+    private final Path snapshotsDirectory;
 
     private final Peers peers;
     private final Local local;
 
-    private static class InstanceHolder
+    /**
+     * Convenience method to retrieve the production code singleton.
+     */
+    public static Nodes nodes()
     {
-        // put into subclass for lazy initialization
-        private static final Nodes instance = new Nodes();
+        return Instance.nodes();
     }
 
     /**
-     * Returns the singleton instance
+     * Convenience method to retrieve the production code singleton.
      */
-    public static Nodes getInstance()
+    public static Local local()
     {
-        return InstanceHolder.instance;
+        return nodes().local;
     }
 
     /**
-     * Returns the singleton instance of {@link Peers}
+     * Convenience method to retrieve the production code singleton.
      */
     public static Peers peers()
     {
-        return getInstance().getPeers();
+        return nodes().peers;
     }
 
-    /**
-     * Returns the singleton instance of {@link Local}
-     */
-    public static Local local()
+    public void shutdown()
     {
-        return getInstance().getLocal();
+        local.close();
+        peers.close();
     }
 
-    /**
-     * Returns information about the node with the given address - if the node address matches the local (broadcast)
-     * address, the returned object is a {@link LocalInfo}. Otherwise, it is {@link PeerInfo} (or {@code null} if no
-     * informatino is available).
-     */
-    @Nullable
-    public static INodeInfo<?> localOrPeerInfo(InetAddressAndPort endpoint)
+    public void syncToDisk()
     {
-        return Objects.equals(endpoint, FBUtilities.getBroadcastAddressAndPort()) ? local().get() : peers().get(endpoint);
+        local.syncToDisk();
+        peers.syncToDisk();
     }
 
-    public static Optional<INodeInfo<?>> localOrPeerInfoOpt(InetAddressAndPort endpoint)
+    public void snapshot(String snapshotName) throws IOException
     {
-        return Optional.ofNullable(localOrPeerInfo(endpoint));
+        syncToDisk();
+
+        // Synchronize to make sure that no saves happen during snapshot creation, which could
+        // make us miss a file or cause spurious errors.
+        synchronized (local)
+        {
+            synchronized (peers)
```

Comment:
I agree with this looking fragile so I've removed these synchronizations and moved them to the `snapshot` methods. We do need to synchronize somewhere because as the comment says we don't want to snapshot in the middle of a write. Also I don't think we need a global lock because, generally, local and peers are looked on as separate entities and there is no real synchronization between them.